git/packet_reader.go

```diff
@@ -0,0 +1,62 @@
+package git
+
+import (
+	"io"
+	"math"
+)
+
+type packetReader struct {
+	proto *protocol
+
+	buf []byte
+}
+
+var _ io.Reader = new(packetReader)
+
+func (r *packetReader) Read(p []byte) (int, error) {
+	var n int
+
+	if len(r.buf) > 0 {
+		// If there is data in the buffer, shift as much out of it and
+		// into the given "p" as we can.
+		n = int(math.Min(float64(len(p)), float64(len(r.buf))))
```

Comment:
Nice catch. Though there aren't any mismatches under MaxPacketLength (see: https://play.golang.org/p/hUhRZhv3n1), I think this is still a good thing to guard against.

I added two functions to the `tools` packet, `MinInt(a, b int) int` and `MaxInt(a, b int) int` in https://github.com/github/git-lfs/pull/1621/commits/7dea0af323d96c60cbd5e508c629a4bb209d01de and used them in both places here.